godot-core/src/registry/plugin.rs

```diff
@@ -57,195 +93,470 @@ impl fmt::Debug for ErasedRegisterRpcsFn {
-        user_property_can_revert_fn: Option<
-            unsafe extern "C" fn(
-                p_instance: sys::GDExtensionClassInstancePtr,
-                p_name: sys::GDExtensionConstStringNamePtr,
-            ) -> sys::GDExtensionBool,
-        >,
-
-        user_property_get_revert_fn: Option<
-            unsafe extern "C" fn(
-                p_instance: sys::GDExtensionClassInstancePtr,
-                p_name: sys::GDExtensionConstStringNamePtr,
-                r_ret: sys::GDExtensionVariantPtr,
-            ) -> sys::GDExtensionBool,
-        >,
-    },
-
-    DynTraitImpl {
-        /// TypeId of the `dyn Trait` object.
-        dyn_trait_typeid: any::TypeId,
-
-        /// Function that converts a `Gd<Object>` to a type-erased `DynGd<Object, dyn Trait>` (with the latter erased for common storage).
-        erased_dynify_fn: fn(Gd<classes::Object>) -> ErasedDynGd,
-    },
+        set(
+            &mut self.user_free_property_list_fn,
+            callbacks::free_property_list::<T>,
+        );
+        self
+    }
+
+    pub fn with_property_get_revert<T: GodotClass + cap::GodotPropertyGetRevert>(mut self) -> Self {
+        set(
+            &mut self.user_property_get_revert_fn,
+            callbacks::property_get_revert::<T>,
+        );
+        set(
+            &mut self.user_property_can_revert_fn,
+            callbacks::property_can_revert::<T>,
+        );
+        self
+    }
 }
 
-pub struct ErasedDynGd {
-    pub boxed: Box<dyn Any>,
+/// Representation of a `#[godot_dyn]` invocation.
+///
+/// Stores all the information needed for `DynGd` re-enrichment.
+#[derive(Clone, Debug)]
+pub struct DynTraitImpl {
+    /// The class that this `dyn Trait` implementation corresponds to.
+    class_name: ClassName,
+
+    /// TypeId of the `dyn Trait` object.
+    dyn_trait_typeid: any::TypeId,
+
+    /// Function used to get a `DynGd<T,D>` from a `Gd<Object>`. This is used in the [`FromGodot`](crate::meta::FromGodot) implementation
+    /// of [`DynGd`]. This function is always implemented as [`callbacks::dynify_fn::<T, D>`] where `T` is the class represented by `class_name`
+    /// and `D` is the trait object corresponding to `dyn_trait_typeid`.
+    ///
+    /// Function that converts a `Gd<Object>` to a type-erased `DynGd<Object, dyn Trait>` (with the latter erased for common storage).
+    erased_dynify_fn: ErasedDynifyFn,
+}
+
+impl DynTraitImpl {
+    pub fn new<T, D>() -> Self
+    where
+        T: GodotClass
+            + Inherits<Object>
+            + crate::obj::AsDyn<D>
+            + Bounds<Declarer = bounds::DeclUser>,
+        D: ?Sized + 'static,
+    {
+        Self {
+            class_name: T::class_name(),
+            dyn_trait_typeid: std::any::TypeId::of::<D>(),
+            erased_dynify_fn: callbacks::dynify_fn::<T, D>,
+        }
+    }
+
+    /// The type id of the trait object this was registered with.
+    pub fn dyn_trait_typeid(&self) -> &any::TypeId {
+        &self.dyn_trait_typeid
+    }
```

Comment:
`TypeId` implements `Copy`. Although [its docs](https://doc.rust-lang.org/std/any/struct.TypeId.html) don't specifically mention that it's "lightweight", we can probably assume that it's likely some sort of integer. Should we return by value, like we do with `Vector4`, `Quaternion` etc?